[feat](spill) spill and reserve

mrhhsg opened the pull request:
### What problem does this PR solve?

Problem Summary:

A brand-new spilling triggering strategy:
1. Use workload groups to control/manage the memory usage of queries. 
2. Trigger spilling when the memory reservation attempt fails.

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

Thearas commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

mrhhsg commented:
run buildall

jacktengg committed:
[fix](spill) improve and fix some bugs

1. improve exchange sink low mem mode;
2. improve exchange memory usage counter
3. fix string hash table memory usage;
4. fix deadlock if disable_memory_gc;
5. improve some log prints.

mrhhsg committed:
[feat](squashed) spill and reserve

squashed from: 54a152d2faffdffc5e2dc150f5d25b69ed30760f to cfb01d4bf0743b7ef74f0421b78bd08cbe501fa3

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

mrhhsg commented:
run buildall

xinyiZzz commented on be/src/common/daemon.cpp:
```diff
@@ -322,15 +322,18 @@ void Daemon::memory_maintenance_thread() {
+        // step 7: handle paused queries(caused by memory insufficient)
+        doris::ExecEnv::GetInstance()->workload_group_mgr()->handle_paused_queries();
+
+        // step 8. Analyze blocking queries.
```
Remove [step 8. Analyze blocking queries.]
The TODO [step 8. Analyze blocking queries.] here and the newly added [step 7: handle paused queries (caused by insufficient memory)] have the same meaning.

mrhhsg commented:
run buildall

xinyiZzz commented on be/src/runtime/memory/thread_mem_tracker_mgr.h:
```diff
@@ -278,40 +279,57 @@ inline void ThreadMemTrackerMgr::flush_untracked_mem() {
 
-inline doris::Status ThreadMemTrackerMgr::try_reserve(int64_t size) {
+inline doris::Status ThreadMemTrackerMgr::try_reserve(int64_t size,
+                                                      bool only_check_process_memory) {
```
为什么 FlushToken::_try_reserve_memory 需要 only_check_process_memory=true 呢

只要让 FlushToken::_try_reserve_memory scoped 的 tracker limit = -1，workload group ptr = null，自然 try reserve 就只检查 process memory 了

现在为了支持 only_check_process_memory，在 memory tracker 和 work load group  分别新增了一个 reserve 方法，代码看着更乱了

yiguolei replied:
@jacktengg 看看

jacktengg replied:
因为memtable fush exectuor 是BE上所有workload group 共享的，在其中reserve的内存也是只算在整个BE process上，不单独计入某个workload group或者查询。
修改 tracker limit = -1或者workload group ptr = null，当前风险较高，先保持现状吧，后续再优化。

jacktengg commented on be/src/olap/memtable_flush_executor.cpp:
```diff
@@ -137,6 +139,39 @@ Status FlushToken::wait() {
+                    PrettyPrinter::print_bytes(size));
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        } else {
+            st = Status::OK();
```
是的，至少保证一个 memtable 可以执行下刷，避免卡死。

jacktengg commented on be/src/olap/memtable_memory_limiter.cpp:
```diff
@@ -110,7 +111,23 @@ int64_t MemTableMemoryLimiter::_need_flush() {
+    // Wait at most 1s, because this code is not aware cancel flag. If the load task is cancelled
+    // Should releae memory quickly.
+    using namespace std::chrono_literals;
+    int32_t sleep_times = 10;
```
是最长等待1s. 每次循环等待100ms，最多循环10次.

jacktengg commented on be/src/olap/memtable_memory_limiter.cpp:
```diff
@@ -243,8 +306,12 @@ void MemTableMemoryLimiter::_refresh_mem_tracker() {
-            _flush_mem_usage += writer->mem_consumption(MemType::FLUSH);
-            _queue_mem_usage += writer->mem_consumption(MemType::WRITE_FINISHED);
+
+            auto flush_usage = writer->mem_consumption(MemType::FLUSH);
```
No.

yiguolei commented on be/src/common/daemon.cpp:
```diff
@@ -547,7 +550,9 @@ void Daemon::cache_adjust_capacity_thread() {
-        double adjust_weighted = GlobalMemoryArbitrator::last_cache_capacity_adjust_weighted;
+        double adjust_weighted = std::min<double>(
+                GlobalMemoryArbitrator::last_cache_capacity_adjust_weighted,
+                GlobalMemoryArbitrator::last_wg_trigger_cache_capacity_adjust_weighted);
```
这里是要直接降低到0.01% 或者说是禁用掉，跟那个递减的逻辑是分开的

xinyiZzz replied:
任何 Query 报错进程内存超限，就把所有 Cache 都清空，感觉有点激进，线上环境可能某个大查询导致所有 Cache 清空后，集群所有任务的耗时增加，QPS降低。

就这样观察下吧，之前讨论过这个问题，目前这样实现好解释

yiguolei commented on be/src/olap/memtable_memory_limiter.cpp:
```diff
@@ -110,7 +111,23 @@ int64_t MemTableMemoryLimiter::_need_flush() {
+    // Should releae memory quickly.
+    using namespace std::chrono_literals;
+    int32_t sleep_times = 10;
+    while (wg != nullptr && wg->enable_write_buffer_limit() && wg->exceed_write_buffer_limit() &&
```
这个enable write buffer limit 只有等到整个wg的内存不足的时候才会动态的设置为true，默认是false。

yiguolei commented on be/src/pipeline/pipeline_task.cpp:
```diff
@@ -340,36 +370,127 @@ Status PipelineTask::execute(bool* eos) {
             _get_block_counter->update(1);
-            RETURN_IF_ERROR(_root->get_block_after_projects(_state, block, eos));
-        }
+            const auto reserve_size = _root->get_reserve_mem_size(_state);
```
root 是pipeline 中负责流式计算的部分，他有可能需要比较大的内存，比如一个宽列的时候，join 产生一次就需要比较大的内存。

yiguolei commented on be/src/runtime/memory/global_memory_arbitrator.h:
```diff
@@ -149,15 +149,16 @@ class GlobalMemoryArbitrator {
-                "{} exceed limit {} or {} less than low water mark {}", process_memory_used_str(),
-                MemInfo::mem_limit_str(), sys_mem_available_str(),
+                "{} exceed limit {} or {} less than low water mark {}",
+                process_memory_used_details_str(), MemInfo::mem_limit_str(),
```
我们先输出，等spill 稳定了，在改小吧

xinyiZzz replied:
TODO: 
打details后，检查下 mysql client 是否会截断错误信息

yiguolei commented on be/src/runtime/memory/mem_tracker_limiter.cpp:
```diff
@@ -81,6 +81,11 @@ std::shared_ptr<MemTrackerLimiter> MemTrackerLimiter::create_shared(MemTrackerLi
                                                                     int64_t byte_limit) {
     auto tracker = std::make_shared<MemTrackerLimiter>(type, label, byte_limit);
+    // Write tracker is only used to tracker the size, so limit == -1
+    auto write_tracker = std::make_shared<MemTrackerLimiter>(type, "Memtable" + label, -1);
```
现在应该是认为这2个没关系，我们去cancel的时候也认为memtable 的buffer 部分不算到query 里的。

xinyiZzz replied:
TODO: 
MemoryProfile 中应该在 Load Memory 中区分 Framgnet 和 Memtable 内存

yiguolei commented on be/src/runtime/memory/mem_tracker_limiter.h:
```diff
@@ -146,6 +147,7 @@ class MemTrackerLimiter final {
     // Log the memory usage when memory limit is exceeded.
     std::string tracker_limit_exceeded_str();
     bool is_overcommit_tracker() const { return type() == Type::QUERY || type() == Type::LOAD; }
+    void set_limit(int64_t new_mem_limit) { _limit = new_mem_limit; }
```
是workload group 运行过程中动态调整的，否则没法做query limit的软限

yiguolei commented on be/src/runtime/memory/mem_tracker_limiter.h:
```diff
@@ -350,9 +369,16 @@ inline void MemTrackerLimiter::cache_consume(int64_t bytes) {
 inline Status MemTrackerLimiter::check_limit(int64_t bytes) {
-    if (bytes <= 0 || (is_overcommit_tracker() && config::enable_query_memory_overcommit)) {
+    // Do not enable check limit, because reserve process will check it.
+    if (bytes <= 0 || _enable_reserve_memory) {
```
现在估计不太行，因为reserve 计算不准，这里得依靠hard limit 来gc

xinyiZzz replied:
所以如果 reserve 成功，那后续 operator 申请内存预期不应在 Allocator 中 check query limit 报错对吧，即使实际申请的内存超过了 reserve 的值。

因为 reserve 失败可以执行 spill，但 Allocator 中报错就直接 cancel query 了，所以除非 hard limit 了，否则不应 cancel query

yiguolei commented on be/src/runtime/memory/memory_profile.cpp:
```diff
@@ -334,15 +333,18 @@ int64_t MemoryProfile::other_current_usage() {
 }
 
+std::string MemoryProfile::process_memory_detail_str() const {
+    return fmt::format("Process Memory Summary: {}\n, {}\n, {}\n, {}\n, {}\n, {}\n",
```
@jacktengg 改一下

yiguolei commented on be/src/runtime/memory/mem_tracker_limiter.cpp:
```diff
@@ -81,6 +81,11 @@ std::shared_ptr<MemTrackerLimiter> MemTrackerLimiter::create_shared(MemTrackerLi
                                                                     int64_t byte_limit) {
     auto tracker = std::make_shared<MemTrackerLimiter>(type, label, byte_limit);
+    // Write tracker is only used to tracker the size, so limit == -1
+    auto write_tracker = std::make_shared<MemTrackerLimiter>(type, "Memtable" + label, -1);
```
@jacktengg jack改一下

yiguolei commented on be/src/runtime/memory/memory_profile.cpp:
```diff
@@ -154,8 +154,7 @@ void MemoryProfile::refresh_memory_overview_profile() {
     ExecEnv::GetInstance()->segments_no_cache_mem_tracker()->set_consumption(
-            MetadataAdder<segment_v2::Segment>::get_all_segments_estimate_size() -
-            SegmentLoader::instance()->cache_mem_usage());
+            MetadataAdder<segment_v2::Segment>::get_all_segments_size());
```
@jacktengg 看看

mrhhsg commented:
run buildall

jacktengg commented:
run buildall

jacktengg commented:
run performance

jacktengg committed:
fix be ut failure

jacktengg commented:
run buildall

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+                // need to check config::disable_memory_gc here, if not, when config::disable_memory_gc == true,
+                // cache is not adjusted, query_it->cache_ratio_ will always be 1, and this if branch will nenver
+                // execute, this query will never be resumed, and will deadlock here
+                if ((!config::disable_memory_gc && query_it->cache_ratio_ < 0.05) ||
```
实际上 cache 释放需要时间，几十G的 page cache 全部释放需要接近1s，

这里 `cache_ratio_ < 0.05` 期望的语义是 "当前所有Cache已经都释放完了，如果其他地方不能释放内存就 spill"，但实际上 Cache 可能并没有释放完。

TODO：给 CacheManager 加个方法，返回所有 Cache 的使用率，若低于某个值，则认为 "所有Cache释放完了，不必等待Cache释放内存了"

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+
+// 1. Sort all memory limiter in all overcommit wg, and cancel the top usage task that with most memory.
+// 2. Maybe not valid because it's memory not exceed limit.
+int64_t WorkloadGroupMgr::cancel_top_query_in_overcommit_group_(int64_t need_free_mem,
```
TODO：我后面会实现，但如上所说，revoke overcommited wg 的触发位置有待商榷

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+    }
+    if (hard_limit) {
+        freed_mem = cancel_top_query_in_overcommit_group_(need_free_mem - total_freed_mem,
+                                                          doris::QUERY_MIN_MEMORY, profile.get());
```
这里为啥要判断 hard_limit，然后设置一个 cancel 的 `lower_bound` 呢

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+    return 0;
+}
+
+int64_t WorkloadGroupMgr::revoke_memory_from_other_group_(std::shared_ptr<QueryContext> requestor,
```
所以支持 overcommit 的 wg 内存 overcommited 后，只能是 process memory 超过 soft limit 后，由其他 wg 的 query 去释放这个 overcommited 的 wg 对吧。

这个调用关系看似合理，但会不会导致频繁调用呢，比如 100个暂停的 query 都去 revoke 其他 wg。
类似的需求，我第一反应在 GC 线程里通过一定策略去 revoke overcommited 的 wg。

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+                                continue;
+                            }
+                        }
+                    } else {
```
如果已经有 query A 去 revoke 了其他 overcommited wg，并将 has_revoked_from_other_group 设为 true

此处让其他 query 直接继续执行不合理，因为 query A  的 reserve size 可能很小，所以 revoke 其他 overcommited wg 很小一块内存。

其他 query 继续执行的语义是 "所有能释放的内存都释放了，所以不管三七二十一全跑起来吧! "，但可能还有 overcommited wg 没有全部释放 overcommited 的部分。

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +257,642 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+                            VLOG_DEBUG << "Query: " << print_id(query_ctx->query_id())
+                                       << " is resumed after revoke memory from other group.";
+                            query_it = queries_list.erase(query_it);
+                            // Do not care if the revoked_size > reserve size, and try to run again.
```
这里不管 revoked_size 是否大于 reserve size，都继续执行 query，真的没问题么。

因为没有让 adjusted_mem_limit 生效，所以 query 重新执行后不会因为 `QUERY_MEMORY_EXCEEDED` 进入暂停状态

Query 再次进入暂停状态大概率还是因为 `process memory exceed`，那就会不断 revoke  overcommited 的 wg，直到 revoke = 0，没有 overcommited 的 wg 后，才会去 spill 当前 query。看似合理，但总感觉线上会有坑，比如 query 看起来`一卡一卡`的

xinyiZzz replied:
预期就是这样，靠轮训解决问题，频繁进入paused queries是预期的

jacktengg commented:
run cloudut

yiguolei reviewed (approved):
LGTM

dataroaring reviewed (approved):
LGTM

xinyiZzz commented on be/src/runtime/query_context.cpp:
```diff
@@ -146,34 +151,45 @@ QueryContext::QueryContext(TUniqueId query_id, ExecEnv* exec_env,
+    }
+    // If the query is a pure load task(streamload, routine load, group commit), then it should not use
+    // memlimit per query to limit their memory usage.
+    if (is_pure_load_task()) {
```
不需要支持限制单个load的memtable

xinyiZzz commented on be/src/runtime/runtime_state.h:
```diff
@@ -605,7 +643,16 @@ class RuntimeState {
+    }
+
+    size_t minimum_operator_memory_required_bytes() const {
+        if (_query_options.__isset.minimum_operator_memory_required_kb) {
```
很难区分哪些算子会用内存，所以都reserver，失败进pause list

xinyiZzz commented on be/src/runtime/workload_group/workload_group_manager.cpp:
```diff
@@ -287,6 +256,624 @@ void WorkloadGroupMgr::refresh_workload_group_metrics() {
+                }
+            } else if (query_ctx->paused_reason().is<ErrorCode::WORKLOAD_GROUP_MEMORY_EXCEEDED>()) {
+                // Only deal with non overcommit workload group.
+                if (wg->enable_memory_overcommit()) {
```
不会，disable 后就不会再进到paused list里了，所以不会死等，如果内存不足就被 allocator或gc cancel。
